src/hooks/UseGraph/useGraph.ts

```diff
@@ -0,0 +1,392 @@
+   *
+   * @type {MapAtCoordinates<TData, TDimensions>}
+   */
+  updateAtCoordinates: UpdateAtCoordinates<TData, TDimensions>
+  updateVertex: UpdateVertex<TData, TDimensions>
+
+  clear: Clear
+  pruneAtCoordinates: PruneAtCoordinates<TDimensions>
+  pruneVertex: PruneVertex<TDimensions>
+
+  map: Map<TData, TDimensions>
+  /**
+   * Returns a TData transformed into a TResult
+   *
+   * @type {MapAtCoordinates<TData, TDimensions>}
+   */
+  mapAtCoordinates: MapAtCoordinates<TData, TDimensions>
+  mapVertex: MapVertex<TData, TDimensions>
+
+  /**
+   * Executes a callback on all vertices of the graph
+   *
+   * @type {ForEachVertex<TData, TDimensions>}
+   */
+  forEachVertex: ForEachVertex<TData, TDimensions>
+  /**
+   * Returns `true` if any vertex returns `true` for the `callback`
+   *
+   * @type {SomeVertex<TData, TDimensions>}
+   */
+  someVertex: SomeVertex<TData, TDimensions>
+  /**
+   * Sets all vertices of the graph to a new value
+   *
+   * @type {SetAllVertices<TData>}
+   */
+  setAllVertices: SetAllVertices<TData>
+  /**
+   * Transforms all vertices of the graph
+   *
+   * @type {UpdateAllVertices<TData, TDimensions>}
+   */
+  updateAllVertices: UpdateAllVertices<TData, TDimensions>
+  /**
+   * Creates a new graph with TData transformed to TResult
+   *
+   * @type {MapAllVertices<TData, TDimensions>}
+   */
+  mapAllVertices: MapAllVertices<TData, TDimensions>
+}
+
+/**
+ * The type of the arguments to `useGraph`
+ *
+ * @export
+ * @typedef {UseGraphOptions}
+ * @template TData
+ * @template {number} [TDimensions=0]
+ */
+export interface UseGraphOptions<TData, TDimensions extends number = 0> {
+  dimensions?: TDimensions
+  initial?: Graph<TData, TDimensions>
+}
+
+/**
+ * The return type for the `useGraph` hook
+ *
+ * @typedef {UseGraphReturn}
+ * @template TData
+ * @template {number} [TDimensions=0]
+ */
+export type UseGraphReturn<TData, TDimensions extends number = 0> = [
+  data: GraphData<TData, TDimensions> | null,
+  UseGraphReturnFunctions<TData, TDimensions>,
+]
+
+/**
+ * The type of the `useGraph` hook
+ *
+ * @typedef {UseGraph}
+ */
+export type UseGraph = <TData, TDimensions extends number = 0>(
+  options?: UseGraphOptions<TData, TDimensions>,
+) => UseGraphReturn<TData, TDimensions>
+
+/**
+ * A hook for managing state with any number of dimensions
+ *
+ * @template TData
+ * @template {number} [TDimensions=0]
+ * @param {UseGraphOptions<TData, TDimensions>} [param0={ dimensions: 0 }]
+ * @param {TDimensions} param0.dimensions
+ * @param {Graph<TData, TDimensions>} param0.initial
+ * @returns {UseGraphReturn<TData, TDimensions>}
+ */
+const useGraph: UseGraph = <TData, TDimensions extends number = 0>(
+  {
+    dimensions,
+    initial,
+    // @ts-expect-error TDimensions defaults to 0 if dimensions is not defined, so this is okay
+  }: UseGraphOptions<TData, TDimensions> = { dimensions: 0 },
+): UseGraphReturn<TData, TDimensions> => {
```

Comment:
Same as a comment elsewhere

```suggestion
const useGraph: UseGraph = <TData, TDimensions extends number = 0>({
  dimensions = 0 as TDimensions,
  initial,
}: UseGraphOptions<TData, TDimensions> = {}): UseGraphReturn<TData, TDimensions> => {
```